core/domain/wipeout_service_test.py

```diff
@@ -63,6 +64,40 @@
 datastore_services = models.Registry.import_datastore_services()
 
 
+def empty_function(*_):
+    """Empty function to mock
+    mailchimp_services.permanently_delete_user_from_list which consists of just
+    an API call which is separately tested.
+    """
+    pass
+
+
+def swapped_add_or_update_function(
+        user_id, _user_email, can_receive_email_updates):
+    """Mock function that just updates the UserPreferencesModel without
+    calling the mailchimp api for testing.
+
+    Args:
+        user_id: str. The ID of the user.
+        _user_email: str. Email ID of the user.
+        can_receive_email_updates: bool. Whether the user can receive
+            email updates.
+    """
+    email_preferences_model = user_models.UserEmailPreferencesModel.get(
+        user_id, strict=False)
+    email_preferences_model.site_updates = can_receive_email_updates
+    email_preferences_model.update_timestamps()
+    email_preferences_model.put()
+
+
+setattr(
+    mailchimp_services, 'permanently_delete_user_from_list',
+    empty_function)
+setattr(
+    mailchimp_services, 'add_or_update_mailchimp_user_status',
+    swapped_add_or_update_function)
```

Comment:
Not sure if this is fine, but basically `wipeout_service.pre_delete_user` is called throughout this test (~90 times I believe) and that function calls both the mailchimp functions either directly or indirectly. So, the alternative was to add a swap statement  for each of the calls which felt too repetitive.

Reply:
I think the way this is usually handled is to have a test stub similar to ElasticSearchStub in test_utils.py. That just uses a local function and avoids connection with the outside world. For similar reasons we may need to implement a "dev mode" implementation in core/platform/bulk_email -- see what is done for email_manager.

That should address the issue. I think what is done here is hacky and isn't the right way to do it, unfortunately.

Please let me know if you have more questions about this!

Reply:
Done, didn't need to use stubbing, since in dev_mode, it automatically chooses the dev_mode_bulk_email_services function which doesn't reference mailchimp.